File: go-sdk/codec/decode_ix.go
package codec

import (
	"fmt"

	"github.com/ifx-run/ifx/go-sdk/constants"
)

// IfxIxName is the on-chain instruction name for a 1-byte discriminator.
type IfxIxName string

const (
	IxNameCreateFrame  IfxIxName = "ifx_create_frame"
	IxNameCloseFrame   IfxIxName = "ifx_close_frame"
	IxNameResetFrame   IfxIxName = "ifx_reset_frame"
	IxNameLet          IfxIxName = "ifx_let"
	IxNameAssert       IfxIxName = "ifx_assert"
	IxNameAssertMulti  IfxIxName = "ifx_assert_multi"
	IxNamePatchedCpi   IfxIxName = "ifx_patched_cpi"
	IxNameIfElse       IfxIxName = "ifx_if_else"
)

var discToName = map[uint8]IfxIxName{
	constants.IxDiscCreateFrame:  IxNameCreateFrame,
	constants.IxDiscCloseFrame:   IxNameCloseFrame,
	constants.IxDiscResetFrame:   IxNameResetFrame,
	constants.IxDiscLet:          IxNameLet,
	constants.IxDiscAssert:       IxNameAssert,
	constants.IxDiscAssertMulti:  IxNameAssertMulti,
	constants.IxDiscPatchedCpi:   IxNamePatchedCpi,
	constants.IxDiscIfElse:       IxNameIfElse,
}

// DecodedIfxInstruction is the result of DecodeIfxInstruction.
type DecodedIfxInstruction struct {
	Name          IfxIxName
	Discriminator uint8
	Data          []byte
	Payload       []byte
}

// DecodeIfxInstruction reads the 1-byte Ifx discriminator (inspection / debugging only).
func DecodeIfxInstruction(data []byte) (DecodedIfxInstruction, error) {
	if len(data) < 1 {
		return DecodedIfxInstruction{}, fmt.Errorf("ifx instruction data is empty")
	}
	disc := data[0]
	name, ok := discToName[disc]
	if !ok {
		return DecodedIfxInstruction{}, fmt.Errorf("unknown Ifx instruction discriminator: %d", disc)
	}
	return DecodedIfxInstruction{
		Name:          name,
		Discriminator: disc,
		Data:          data,
		Payload:       data[1:],
	}, nil
}

// IfxIxHint returns the instruction name or empty string when unknown.
func IfxIxHint(data []byte) string {
	dec, err := DecodeIfxInstruction(data)
	if err != nil {
		return ""
	}
	return string(dec.Name)
}

File: rust-sdk/src/decode_ix.rs
//! Ifx instruction discriminator decode (inspection / debugging).

use crate::constants::{
    IX_DISC_ASSERT, IX_DISC_ASSERT_MULTI, IX_DISC_CLOSE_FRAME, IX_DISC_CREATE_FRAME,
    IX_DISC_IF_ELSE, IX_DISC_LET, IX_DISC_PATCHED_CPI, IX_DISC_RESET_FRAME,
};
use crate::error::ScratchError;

/// On-chain Ifx instruction name for a 1-byte discriminator.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum IfxIxName {
    CreateFrame,
    CloseFrame,
    ResetFrame,
    Let,
    Assert,
    AssertMulti,
    PatchedCpi,
    IfElse,
}

impl IfxIxName {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::CreateFrame => "ifx_create_frame",
            Self::CloseFrame => "ifx_close_frame",
            Self::ResetFrame => "ifx_reset_frame",
            Self::Let => "ifx_let",
            Self::Assert => "ifx_assert",
            Self::AssertMulti => "ifx_assert_multi",
            Self::PatchedCpi => "ifx_patched_cpi",
            Self::IfElse => "ifx_if_else",
        }
    }
}

/// Decoded Ifx instruction header (args not fully deserialized).
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct DecodedIfxInstruction {
    pub name: IfxIxName,
    pub discriminator: u8,
    pub data: Vec<u8>,
    pub payload: Vec<u8>,
}

pub fn decode_ifx_instruction(data: &[u8]) -> Result<DecodedIfxInstruction, ScratchError> {
    let disc = *data.first().ok_or_else(|| ScratchError::Encode("ifx instruction data is empty".into()))?;
    let name = match disc {
        IX_DISC_CREATE_FRAME => IfxIxName::CreateFrame,
        IX_DISC_CLOSE_FRAME => IfxIxName::CloseFrame,
        IX_DISC_RESET_FRAME => IfxIxName::ResetFrame,
        IX_DISC_LET => IfxIxName::Let,
        IX_DISC_ASSERT => IfxIxName::Assert,
        IX_DISC_ASSERT_MULTI => IfxIxName::AssertMulti,
        IX_DISC_PATCHED_CPI => IfxIxName::PatchedCpi,
        IX_DISC_IF_ELSE => IfxIxName::IfElse,
        _ => {
            return Err(ScratchError::Encode(format!(
                "unknown Ifx instruction discriminator: {disc}"
            )))
        }
    };
    Ok(DecodedIfxInstruction {
        name,
        discriminator: disc,
        data: data.to_vec(),
        payload: data[1..].to_vec(),
    })
}

pub fn ifx_ix_hint(data: &[u8]) -> Option<&'static str> {
    decode_ifx_instruction(data)
        .ok()
        .map(|d| d.name.as_str())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn decode_let_ix() {
        let dec = decode_ifx_instruction(&[IX_DISC_LET, 1, 2]).unwrap();
        assert_eq!(dec.name, IfxIxName::Let);
        assert_eq!(dec.payload, &[1, 2]);
        assert_eq!(ifx_ix_hint(&[IX_DISC_RESET_FRAME]), Some("ifx_reset_frame"));
    }
}
